lib/dartdoc.dart

```diff
@@ -44,40 +44,69 @@ class DartdocGeneratorOptionContext extends DartdocOptionContext
       : super(optionSet, dir);
 }
 
+class DartdocFileWriter {
+  final Map<String, Warnable> _writtenFiles = {};
+
+  File write(String filePath, Object content,
+      {bool allowOverwrite, Warnable element}) {
+    allowOverwrite ??= false;
+    if (!allowOverwrite) {
+      if (_writtenFiles.containsKey(filePath)) {
+        assert(element != null,
+            'Attempted overwrite of ${filePath} without corresponding element');
+        Warnable originalElement = _writtenFiles[filePath];
+        Iterable<Warnable> referredFrom =
+            originalElement != null ? [originalElement] : null;
+        element?.warn(PackageWarning.duplicateFile,
+            message: filePath, referredFrom: referredFrom);
+      }
+    }
+
+    var file = File(filePath);
+    var parent = file.parent;
+    if (!parent.existsSync()) {
+      parent.createSync(recursive: true);
+    }
+
+    if (content is String) {
+      file.writeAsStringSync(content);
+    } else if (content is List<int>) {
+      file.writeAsBytesSync(content);
+    } else {
+      throw ArgumentError.value(
+          content, 'content', '`content` must be `String` or `List<int>`.');
+    }
+    _writtenFiles[filePath] = element;
+    return file;
+  }
+}
+
 /// Generates Dart documentation for all public Dart libraries in the given
 /// directory.
 class Dartdoc extends PackageBuilder {
-  final List<Generator> generators;
+  final Generator generator;
   final Set<String> writtenFiles = Set();
   Directory outputDir;
 
   // Fires when the self checks make progress.
   final StreamController<String> _onCheckProgress =
       StreamController(sync: true);
 
-  Dartdoc._(DartdocOptionContext config, this.generators) : super(config) {
+  Dartdoc._(DartdocOptionContext config, this.generator) : super(config) {
     outputDir = Directory(config.output)..createSync(recursive: true);
-    generators.forEach((g) => g.onFileCreated.listen(logProgress));
   }
 
   /// An asynchronous factory method that builds Dartdoc's file writers
   /// and returns a Dartdoc object with them.
   static Future<Dartdoc> withDefaultGenerators(
       DartdocGeneratorOptionContext config) async {
-    List<Generator> generators = await initHtmlGenerators(config);
-    return Dartdoc._(config, generators);
+    return Dartdoc._(
+        config, await initHtmlGenerator(config, DartdocFileWriter().write));
```

Comment:
This suggests to me that DartdocFileWriter could be a class implementing call() -- this way you could just pass an instance of DartdocFileWriter and dispense with the typedef.